### PR TITLE
Update sprite handling

### PR DESCRIPTION
**Pull request by KaiVolland**
This updates the handling of mapbox sprites to make use of the geostyler [`Sprite`](https://github.com/geostyler/geostyler-style/blob/master/style.ts#L390).

:exclamation: breaking change

### CODE REVIEW

File: src/MapboxStyleParser.ts [hunk 1]
```diff
@@ -281,18 +304,31 @@ export class MapboxStyleParser implements StyleParser<Omit<MbStyle, 'sources'>>
...
     if (!this.spriteBaseUrl) {
       return;
     }
-    // TODO update endpoint as soon as api specification was made
-    let url: string = '/sprites/?';
-    url += 'name=' + spriteName;
-    url += '&baseurl=' + encodeURIComponent(this.spriteBaseUrl);
-    return url;
+
+    const response = await fetch(this.spriteBaseUrl + '.json');
```
Comment (simonseyock) [inline src/MapboxStyleParser.ts:315]: should fetching be the default behavior? I'd say yes, but would it be sensible to provide a way around?
I can imagine there are scenarios where you would not want to fetch, for example if the parser is not run in a browser.

Then a sprite declaration would only make sense if you could provide the necessary meta data via another way, maybe to the constructor?

Comment (jansule) [inline src/MapboxStyleParser.ts:315]: I'm wondering if we can place the fetch request at a better position. Right now, we fetch the spritesheet for every distinct icon, which can lead to quite some requests. Maybe we can check before, if sprites are used and fetch once per `readStyle`/`writeStyle`. Or maybe we can do this when instantiating the parser. I think spritesheets might be quite static.

File: data/styles/line_patternline.ts [hunk 2]
```diff
@@ -29,7 +33,19 @@ const linePatternLine: Style = {
...
+          position: [
+            0,
+            0,
+          ],
+          size: [
+            12,
+            12,
+          ],
+        },
+      },
```
[the remark this answers is not in the document]
Comment (KaiVolland) [reply data/styles/line_patternline.ts:48]: As we only differntiate between 'string' and this more or less complex object i think a type guard is pretty straight forward. I don't see the necessity of adding a 'kind' prop here. We also only have these on the symbolizers.

File: src/MapboxStyleParser.ts [hunk 3]
```diff
@@ -434,27 +475,28 @@ export class MapboxStyleParser implements StyleParser<Omit<MbStyle, 'sources'>>
    * @param layout A Mapbox layout
    * @return A GeoStylerStyle-FillSymbolizer
    */
-  getFillSymbolizerFromMapboxFillLayer(paint: FillPaint, layout: FillLayout): FillSymbolizer {
+  async getFillSymbolizerFromMapboxFillLayer(paint: FillPaint, layout: FillLayout): Promise<FillSymbolizer> {
+    const graphicFill = await this.getPatternOrGradientFromMapboxLayer(paint?.['fill-pattern']);
```
Comment (jansule) [inline src/MapboxStyleParser.ts:479]: See comment above

File: src/MapboxStyleParser.ts [hunk 4]
```diff
@@ -281,18 +304,31 @@ export class MapboxStyleParser implements StyleParser<Omit<MbStyle, 'sources'>>
    * @param spriteName Name of the sprite
    * @return the url that returns the single image
    */
-  getIconImage(spriteName: string): (string|undefined) {
+  async getIconImage(spriteName: string): Promise<Sprite | undefined> {
```
[the remark this answers is not in the document]
Comment (KaiVolland) [reply src/MapboxStyleParser.ts:307]: I think it can be quite complex: https://docs.mapbox.com/style-spec/reference/types/#resolvedimage

I'll add a check for string here.

File: src/MapboxStyleParser.ts (hunk 3 above)
Comment (KaiVolland) [reply src/MapboxStyleParser.ts:479]: I did not really change the code here, so this is not part of this PR. But we might want to add some more complex testdata to see where we have some gaps left.

**Comment by KaiVolland** [comment]
I created an issue for refactoring as coding/reviewing is already getting difficult: https://github.com/geostyler/geostyler-mapbox-parser/issues/309

File: src/MapboxStyleParser.ts (hunk 1 above)
Comment (KaiVolland) [reply src/MapboxStyleParser.ts]: https://github.com/geostyler/geostyler-mapbox-parser/issues/310

File: src/MapboxStyleParser.ts [hunk 5]
```diff
@@ -204,8 +230,16 @@ export class MapboxStyleParser implements StyleParser<Omit<MbStyle, 'sources'>>
     'geostyler:ref': GeoStylerRef;
   };
 
+  private gsMetadata: {
```
Comment (jansule) [inline src/MapboxStyleParser.ts:233]: Can we move this into a separate PR for better code review?

**Review by jansule — COMMENTED** [review]
I'm not sure, if I like moving the metadata to a class member. IMHO, this only introduces not needed state handling, just for the benefit of not passing an additional argument to the methods.

I agree that this parser needs some refactoring, but I don't think that this approach is the solution to it.

File: src/MapboxStyleParser.spec.ts [hunk 6]
```diff
@@ -79,14 +92,20 @@ describe('MapboxStyleParser implements StyleParser', () => {
...
     it('can read a mapbox Line style', async () => {
-      expect.assertions(2);
       const { output: geoStylerStyle } = await styleParser.readStyle(mb_line_simpleline);
       expect(geoStylerStyle).toBeDefined();
       expect(geoStylerStyle).toEqual(line_simpleline);
     });
 
     it('can read a mapbox Line style with fill pattern', async () => {
-      expect.assertions(2);
+      mockFetchResult({
```
[the remark this answers is not in the document]
Comment (KaiVolland) [reply src/MapboxStyleParser.spec.ts:101]: It is only needed for certain tests, so I would say it is not needed.

File: src/MapboxStyleParser.ts [hunk 7]
```diff
@@ -825,10 +902,10 @@ export class MapboxStyleParser implements StyleParser<Omit<MbStyle, 'sources'>>
    * @return The Promise resolving with a GeoStylerStyle-ReadStyleResult
    */
   readStyle(mapboxStyle: MbStyle): Promise<ReadStyleResult> {
-    return new Promise<ReadStyleResult>(resolve => {
+    return new Promise<ReadStyleResult>(async resolve => {
       try {
         const mbStyle = structuredClone(mapboxStyle);
-        const geoStylerStyle: Style = this.mapboxStyleToGeoStylerStyle(mbStyle);
+        const geoStylerStyle: Style = await this.mapboxStyleToGeoStylerStyle(mbStyle);
```
[the remark this answers is not in the document]
Comment (KaiVolland) [reply src/MapboxStyleParser.ts:914]: Gets cleaned / set in `mapboxStyleToGeoStylerStyle`.

File: src/MapboxStyleParser.ts (hunk 5 above)
Comment (KaiVolland) [reply src/MapboxStyleParser.ts:233]: I'd prefer to refactor this in a followup. See #309

**Review by jansule — APPROVED** [review]
